Use KahanSummation for AverageAggregation

BaurzhanSakhariev opened the pull request:
Improved precision for average aggregation for floats and doubles.


`create table nums (value double precision);`

`insert into nums (value) (select * from unnest([10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1,10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1,10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1,10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1,10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1,10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1,10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1,10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1,10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1,10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1, 10000.1]));`

`select avg(value) avg from nums; `

Before:

```
cr> select avg(value) avg from nums;                                                                                                                                                                                    
+--------------------+
|                avg |
+--------------------+
| 10000.100000000002 |
+--------------------+
SELECT 1 row in set (0.003 sec)
```

After

```
 select avg(value) avg from nums;                                                                                                                                                                                                   
+---------+
|     avg |
+---------+
| 10000.1 |
+---------+
SELECT 1 row in set (0.035 sec)
```

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

mfussenegger commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/AverageAggregation.java:
```diff
@@ -304,8 +318,7 @@ public Signature boundSignature() {
                         return new AverageState();
                     },
                     (values, state) -> {
-                        state.sum += values.nextValue();
-                        state.count++;
+                        state.addNumber(values.nextValue()); // Mutates state.
```
We should probably follow up here and have a version where `sum` is a `long`, to avoid the kahan summation overhead for non-floating numbers.

BaurzhanSakhariev replied:
I refactored a little bit to use different state for integral numbers - something similar to HllState but it doesn't store dateType as we already want to reduce fields in HllState and I created a package for average related classes - was a bit hard to scroll over current AverageAggregation as it was already big and with new changes similar to HllAggregation (type aware implementation) it became huge.

mfussenegger commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/average/AverageState.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        // Similar to HllState but in order not to store dataType in state we provide minimal information for later reading from StreamInput.
+        out.writeBoolean(this instanceof IntegralAverageState);
+        out.writeDouble(this.sum);
+        out.writeVLong(this.count);
+    }
```
This would break streaming BWC (and therefore break rolling upgrades).

We'll need to add version checks. `newState` of aggregation function receives the `Version minNodeInCluster` property which can be used. We'll have to use the old average state for if the minNodeInCluster is < 4.7.0. To ensure all nodes in a cluster can handle the formats.

I'd actually try to avoid streaming this information altogether. Given that the average aggregation has the type information it should be possible to re-create the right state without streaming this information. (This would mean having 2 type implementations as well)

mfussenegger replied:
This is still not addressed?

BaurzhanSakhariev replied:
oops, I forgot to delete it from here.

mfussenegger commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/average/IntegralAverageState.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Override
+    public void addNumber(Number number) {
+        this.sum += number.doubleValue();
+        this.count++;
```
Since we split up the two states, I think we could also make `sum` a `long` to switch from floating point addition instructions to `ladd`

BaurzhanSakhariev replied:
it will be changed in `OverflowAwareMutableLong` in #11612 - neither double nor long are capable to handle case of 2 `MAX_LONG`-s as nominator will overflow/lose precision. `OverflowAwareMutableLong` would extend `Number` and `double sum`  would be changed to `Number sum`. If we don't want to change return type in 4.7 (at least `bigint` as `avg` belongs to type range and cannot exceed `bigint` for integrals or even numeric to align with PG) then it doesn't make sense to exchange one issue to another.

mfussenegger replied:
I'm thinking we could handle this similarly to how we handled sum, where we have a dedicated NumericSumAggregation and where we require opt-in from the user via explicit cast to numeric to use it. This would avoid the change in behavior and we could optimize the "sum some int values" cases to use integer arithmetic instead of floating arithmetic.

BaurzhanSakhariev replied:
The thing is final average value is not supposed to overflow, it's intermediate nominator sum being the issue - wouldn't it be annoying to users to cast? I think we can go for explicit cast opt-in as a current fix (and add it to docs like we did for sum) but probably would need a follow up in 5.0 where we can afford having breaking change of return type.

If we go with explicit cast opt-in question what do we do in case of overflow - throw ArithmeticException or just warn user that they must cast otherwise they got incorrect result. I feel like throwing exception is more honest than providing wrong result but then again it's behavioural change.

mfussenegger replied:
Not sure if the overflowing behavior of double is better than that of long and this PR already introduces the `IntegralAverageState`, but fine for me to address this in a follow up

mfussenegger commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/average/NonIntegralAverageState.java:
```diff
@@ -0,0 +1,49 @@
+
+import javax.annotation.Nonnull;
+
+public class NonIntegralAverageState extends AverageState {
```
```suggestion
public class FractionalAverageState extends AverageState {
```

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/average/AverageAggregation.java:
```diff
@@ -81,125 +76,21 @@ public static void register(AggregationImplModule mod) {
                         DataTypes.DOUBLE.getTypeSignature()),
-                    AverageAggregation::new
+                    (signature, boundSignature) ->
+                        new AverageAggregation(signature, boundSignature, supportedType)
```
Yep, I'm using `signature.getArgumentDataTypes().get(0)` now, thx.

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/average/AverageAggregation.java:
```diff
@@ -81,125 +76,21 @@ public static void register(AggregationImplModule mod) {
         }
     }
 
-    public static class AverageState implements Comparable<AverageState> {
```
My second commit " Don't use KahanSummation for integral numbers" is mostly about extraction  - but I see your point, I created a commit for extraction but it's a bit synthetic, not really good as if I did it on the first place.

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/average/AverageState.java:
```diff
@@ -0,0 +1,118 @@
+            default:
+                throw new IllegalArgumentException(String.format(Locale.ENGLISH, "data type %s is not supported", dataType.getName()));
+        }
+    }
```
There is one place where I used it and I moved switch there - if I got comment correctly(remove type handing from AverageState). I didn't create a supplier since it would be same amount of code and it's used only in `public AverageState newState`

mfussenegger replied:
My thought was that the registration is already concrete anyways

```
            for (var supportedType : SUPPORTED_TYPES) {
                mod.register(
                    Signature.aggregate(
                        functionName,
                        supportedType.getTypeSignature(),
                        DataTypes.DOUBLE.getTypeSignature()),
                    AverageAggregation::new
                );
            }
```

We could split this into fractional/integral and already provide the right type factory instead of asking about the type instance after the fact.

BaurzhanSakhariev replied:
got it, did one more commit

mfussenegger commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/average/AverageAggregation.java:
```diff
@@ -58,148 +55,58 @@
 
-    static final List<DataType<?>> SUPPORTED_TYPES = Lists2.concat(
-        DataTypes.NUMERIC_PRIMITIVE_TYPES, DataTypes.TIMESTAMPZ);
+    static final List<DataType<?>> SUPPORTED_INTEGRAL_TYPES = List.of(DataTypes.TIMESTAMPZ, DataTypes.BYTE, DataTypes.SHORT, DataTypes.INTEGER, DataTypes.LONG);
+    static final List<DataType<?>> SUPPORTED_FRACTIONAL_TYPES = List.of(DataTypes.FLOAT, DataTypes.DOUBLE);
```
```suggestion
    static final List<DataType<?>> INTEGRAL_TYPES = List.of(DataTypes.TIMESTAMPZ, DataTypes.BYTE, DataTypes.SHORT, DataTypes.INTEGER, DataTypes.LONG);
    static final List<DataType<?>> FRACTIONAL_TYPES = List.of(DataTypes.FLOAT, DataTypes.DOUBLE);
```

Not sure if the `SUPPORTED_` prefix adds much here. There isn't a `UNSUPPORTED_INTEGRAL_TYPES` list, so the distinction/clarification seems unnecessary.

mfussenegger commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/average/AverageAggregation.java:
```diff
@@ -58,148 +55,58 @@
+    private final Supplier<FixedWidthType> stateDateTypeSupplier;
 
-    AverageAggregation(Signature signature, Signature boundSignature) {
+    AverageAggregation(Signature signature, Signature boundSignature, Supplier<AverageState> stateSupplier, Supplier<FixedWidthType> stateDateTypeSupplier) {
```
Given that we have singleton instances for the type you could avoid wrapping it into a `Supplier`. Maybe also make it more concrete - given that there is a dependency that the type value must match the `AverageState`

mfussenegger commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/average/AverageAggregation.java:
```diff
@@ -260,13 +164,13 @@ public AverageState newState(RamAccounting ramAccounting,
     @Override
     public DataType<?> partialType() {
-        return AverageStateType.INSTANCE;
+        return (DataType<?>) stateDateTypeSupplier.get();
```
See previous comment about making the type of the `DateTypeSupplier` more concrete. There shouldn't be a reason to have a cast here.

BaurzhanSakhariev replied:
Done. I created parent class for both types as I still needed cast to FixedWidthType when I did `ramAccounting.addBytes(((FixedWidthType) stateDateType).fixedSize());`

mfussenegger commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/average/FractionalAverageStateType.java:
```diff
@@ -0,0 +1,63 @@
+
+public class FractionalAverageStateType extends AverageStateType {
+
+    static final int ID = 1026;
```
I'm afraid this will still break rolling upgrade compatibility. Old nodes won't be able to handle this new version id.

We'd to ensure that only the previous state type is used if one of the nodes is older than the current version.

mfussenegger replied:
Alternative solution would be to parameterize the add function and keep a single state - given that the internal representation is the same in both cases. 

The current diff may be too influenced by the planned upcoming change to address the overflow handling.

BaurzhanSakhariev replied:
I went with version check approach - was indeed influenced by keeping overflow issue in mind. I swapped id-s of integral and fractional types - now fractional is 1024 (like old averageStateType) and it's BWC with it. And integralType now has 1026 and will be used for with OverflowAwareMutableLong later - it's compatible now but it definitely won't be compatible later

BaurzhanSakhariev replied:
I will revert part of my changes related to future overflow handling and get back to single average state to keep BWC, that would simplify my current PR and to handle precision I will add special numericavgagregation like we did for sum (with hint in docs) - it wouldn't break BWC as new argument -> return type is an extension.

BaurzhanSakhariev commented:
Hi @seut , PR has a lot of comments but I reverted it to the state when I got approval and one [comment](https://github.com/crate/crate/pull/11616#discussion_r680795050) - I addressed it with [this commit](https://github.com/crate/crate/pull/11616/commits/6ac73f5ba994b09210704adb849d3e3ddc9ef5a8) - other comments are outdated and can be skipped.

mfussenegger commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/AverageAggregation.java:
```diff
@@ -100,6 +104,21 @@ public Double value() {
             }
         }
 
+        public void addNumber(double number, boolean isIntegral) {
+            this.sum = isIntegral ? this.sum + number : kahanSummationForDouble.sum(this.sum, number);
+            this.count++;
+        }
```
Sorry I was probably too vague in my last comment. I meant that it could parameterize the functions itself. 

Not sure if it has any performance impact, but it could avoid the branching.

BaurzhanSakhariev replied:
I did another implementation (third commit), here is benchmark results:

current implementation (with branching)

```
  Statement:
     select avg("adRevenue") from uservisits
   Concurrency: 1
   Iterations: 500
100%| 500/500 [00:08<00:00, 58.03 requests/s]
Runtime (in ms):
    mean:    16.172 ± 2.325
    min/max: 13.883 → 606.535
Percentile:
    50:   14.773 ± 26.522 (stdev)
    95:   16.126
    99.9: 606.535
```
    
  Third commit (without branching)
```
  Statement:
     select avg("adRevenue") from uservisits
   Concurrency: 1
   Iterations: 500
100%| 500/500 [00:09<00:00, 50.30 requests/s]
Runtime (in ms):
    mean:    18.553 ± 2.414
    min/max: 15.098 → 628.538
Percentile:
    50:   16.483 ± 27.535 (stdev)
    95:   21.320
    99.9: 628.538
```

It got worse, probably because of extra unboxing in Consumers.

I didn't do smth like `addNumber(double num)` and `addNumber(long num)` since this PR keeps double sum for the sake of BWC (overflow addressed in another PR with numeric type) and I anyway have to do `Number.doubleValue`. another reason for having consumers rather than 2 state changers is that reduce accepts AverageState and I have to do branching there anyway.

mfussenegger replied:
Why don't you use an interface that doesn't require boxing?

BaurzhanSakhariev replied:
I tried FunctionalINterface with primitive argument instead of Consumer (if I got suggestion correctly) but it got even worth

```
 Statement:
     select avg("adRevenue") from uservisits
   Concurrency: 1
   Iterations: 500
100%| 500/500 [00:12<00:00, 40.21 requests/s]
Runtime (in ms):
    mean:    23.574 ± 2.280
    min/max: 19.538 → 597.614
Percentile:
    50:   21.541 ± 26.012 (stdev)
    95:   26.650
    99.9: 597.614
```
    
It looks like some predefined caller (Consumer or Interface) --> call relevant AverageState method is slower than if + direct relevant method call

mfussenegger replied:
Thanks for investigating. Let's go with the initial variant then.

BaurzhanSakhariev replied:
👌
